### ISSUE TITLE
How to automate dependabot updates generating a changeset

### ISSUE THREAD

**Issue by jasikpark** [atlassian/changesets#647]
## Affected Packages

cli/actions, etc.

## Problem

I want to automate my dependency updates w/ dependabot, but I've realized I'm not generating changesets for each dep upgrade - I'd rather sort out a way to automate that, but I'm not sure how!

## Proposed solution

My initial idea would be to have a github action that runs on push to main that looks for whether the commit is tagged w/ "dependabot" and then adds a changeset w/ the content of the commit message?
ex: https://github.com/snowpackjs/prettier-plugin-astro/pull/14

**Comment by jasikpark** [comment]
I see https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/automating-dependabot-with-github-actions, though I imagine I'd need to wait until after the PR to generate the changeset, rather than being able to alter the PR?

**Comment by ReDrUm** [comment]
+1 for this for the same reason (Renovate Bot in my case).

I'm aware I can generate the markdown files myself, but it'd be nice to lean on the existing logic the CLI provides since it already knows which packages have been touched. Something like `yarn changeset add --type patch --message 'Bump eslint from 7.3.2 to 8.0.0' --include-changed` would be pretty powerful IMO.

**Comment by Andarist** [comment]
Ye, the proposed command sounds quite nice - we could add support for it. Although from the Changeset PoV a "changed" package (at least at the moment) is a one that had any files changed since the "base branch". With Renovate bot and stuff you might only change `devDependencies` of a package and such a change most likely doesn't deserve a changeset as it doesn't affect the final consumer of your package.

So I'm unsure what's the best strategy to handle this case - but we could start with a PR that would implement this using the current "changed" logic and bikeshed this with a PR already in place.

I won't have time to work on this right now though so I would appreciate a PR from a community member who cares about this feature :wink: I will get to this eventually but it can take some time for me.

**Comment by jasikpark** [comment]
Right, it definitely seems like a case where you would want to filter and only create changesets for dependency updates and not devDependency updates

**Comment by Andarist** [comment]
Yep, this kinda means that this would be a new kind of a filter for Changesets as no other thing in Changesets check a thing like this - but it seems useful, so maybe it's worth adding support for this. A slight complication is that we support `pnpm` and `pnpm` supports package manifests in different formats than JSON. IIRC they support both `package.json5` and `package.yml` - so we would have to add support for those too and that comes with including parsers for those formats etc. It could be nice if `pnpm` would support smth like `pnpm get-manifest` so we could only invoke their CLI to get this kind of information. OTOH this would involve reading the "before" state to compare both versions, so a solution like this would involve some git shenanigans. So maybe a parser + checking the diff and trying to figure out which part of the file has changed would be better after all. Luckily either format of package manifests shouldnt be hard to parse to assess this.

Anyway - I would still start with a basic functionality, and only then think about adding `--ignore-dev-dep-updates` (or similar).

**Comment by RichiCoder1** [comment]
It's been a while since this issue has been commented on, so I wanted to add a comment.

Is there anyone still to automate this today? I'd love to be able to add automated changsets for renovate/dependabot upgrades, and then be able to merge the release PR.

**Comment by micha149** [comment]
Just giving changesets a first serious try as I am tired of semantic release. To reduce maintenance effort I am also using dependabot and want to automate creating a changeset for its updates.

Now I am realizing the `add` command is not able to receive any of the prompted values as an argument. Is there something planned to make this possible?

**Comment by feelepxyz** [comment]
👋  I've created a basic action to try and automate changesets for dependabot prs: https://github.com/feelepxyz/dependabot-changesets

Would love any early feedback on this! It's definitely still a bit rough around the edges and currently has no tests (coming soon) 🙈 

I started out thinking it would be a five minute job to programmatically call `changesets add` but as @micha149 also found out, it's not possible. After spending some time on this I know get why. The base case where you only have one package per repo would be very easy to support, but changesets supports monorepos with potentially hundreds of packages. 

A single dependabot pull request might update several packages, so we need to map updated packages to updated dependencies and create a changelog for each one. 

I opted to create empty changelogs for transitive/dev dependency updates, and top-level updates that only update  `package-lock.json` as these should not affect downstream consumers of your package. 

Is this the right thing to do? Are there cases where you would like a changeset for these types of updates? 🤔

**Comment by psteinroe** [comment]
hey everyone, I am using a copy of [this](https://github.com/backstage/backstage/blob/master/.github/workflows/sync_renovate-changesets.yml) workflow to auto-generate change sets for renovate prs. It works well! Note that you have to change the if statement, and remove the token.

**Comment by Enalmada** [comment]
FYI It looks like @scaleway has created an npm module [@scaleway/changesets-renovate](https://www.npmjs.com/package/@scaleway/changesets-renovate) based on the above workflow.  

You can see it in action in their [.github/workflows/changesets-renovate.yml](https://github.com/scaleway/scaleway-lib/blob/main/.github/workflows/changesets-renovate.yml)

I tried it out and it seems to be working:
<img width="692" alt="image" src="https://github.com/changesets/changesets/assets/1892132/8084ab1c-88a0-499c-871e-8711a1bc3074">

I made a shared workflow out of it so I could reuse the same code in a bunch of npm library repositories.

parent repo:
```
name: Generate changeset for Renovate

on:
  workflow_call:

jobs:
  generate-changeset:
    runs-on: ubuntu-latest
    if: github.actor == 'renovate[bot]'
    steps:
      - name: Checkout Repo
        uses: actions/checkout@v4
        with:
          fetch-depth: 2
          ref: ${{ github.head_ref }}

      - name: Setup Bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest

      - name: Install Dependencies
        run: bun install

      - name: Git Identity
        run: |
          git config --global user.name 'Renovate Bot'
          git config --global user.email 'noreply@mend.io'

      - name: Run changesets-renovate
        run: bunx @scaleway/changesets-renovate
```

in each npm module repo:
```
name: Generate changeset for Renovate

on:
  pull_request_target:
    paths:
      - '.github/workflows/changesets-renovate.yml'
      - 'bun.lockb'
      - 'package.json'

jobs:
  generate-changeset:
    uses: enalmada/.github/.github/workflows/changesets-renovate.yml@main
```

**Comment by mscharley** [comment]
I haven't done an extensive audit of this, but be aware that this is setting off alarm bells to me. I would be wary of any solution to this using an npm module, especially in conjunction with the `pull_request_target` trigger. If you haven't read it before, then the security article by Github is a good read: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

I didn't post it to this thread yet because I haven't tested it with dependabot yet, but for renovate specifically I would recommend something like the action I wrote: https://github.com/mscharley/dependency-changesets-action#readme

It is more obviously secure because it doesn't rely on a checkout of the repo, it just uses Github's API to analyse the PR diff and push a new commit to the end of the branch if necessary.

**Comment by aperkaz** [comment]
We use a variation of the following script: https://github.com/backstage/backstage/blob/master/.github/workflows/sync_renovate-changesets.yml

**Comment by guesant** [comment]
by using StafflinePeoplePlus/dependabot-changesets, the default provided example throws codeql alert because of the if label check and checkout to ref 

![Image](https://github.com/user-attachments/assets/5a81f86e-448e-4291-898d-5677958fedbf)

**Comment by mscharley** [comment]
> by using StafflinePeoplePlus/dependabot-changesets, the default provided example throws codeql alert because of the if label check and checkout to ref

Checkout [my project](https://github.com/mscharley/dependency-changesets-action#readme) which doesn't need a checkout at all to add it's changesets which avoids a whole slew of security issues. I don't personally use dependabot but I know that there are at least some people using it with my action now. If you do get it working, I'd very much appreciate some help writing some docs to help onboard people using dependabot with the action.

**Comment by guesant** [comment]
> personally use dependabot but I know that there are at least some people using it with my action now. If you

wow, thank you so much! seems good to me, everything i needed :)

**Comment by guesant** [comment]
After some usage, I found that `mscharley/dependency-changesets-action` seems very good for me. 

Thank you for creating this useful, well maintained and with such code quality @mscharley !

I'm gonna share my experience. 

![Image](https://github.com/user-attachments/assets/f645fa97-20bc-4cdf-a64b-a1237879ed50)


> [!IMPORTANT] 
> There only consideration that I want to particularly discuss is about the `token` being used in the action. The most ease to use is `GITHUB_TOKEN`, but this don't trigger other PR lint workflows; `PAT` seems too personal for me, and may permit read other private repositories if there is some leak; `Github App Token` is the one that I prefer in organizations.

**I personally recommends the `3 - using Github App Token` approach!**

> [!CAUTION] 
> Please read this security guide to how to properly secure your repo againts pwn requests
> https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/
> 
> also NEVER allows that workflows runs for entrusted users and use security scanners like CodeQL, Sonar etc

## 1 - using GITHUB_TOKEN

This is the minimal code that you can use for the automated commit with changesets for affected packages. 

The only trade-off in this approach is that "git push" that uses GITHUB_TOKEN don't trigger other workflows. So, the merge button will never be enabled if you block pull requests that don't passes your lint workflow (see https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/triggering-a-workflow#triggering-a-workflow-from-a-workflow)

```yml
name: Add changeset to Dependabot updates

on:
  pull_request_target:
    types: [opened, synchronize, labeled]

jobs:
  dependabot:
    name: Update Dependabot PR
    if: contains(github.event.pull_request.labels.*.name, 'dependencies')

    runs-on: ubuntu-latest

    permissions:
      contents: write

    steps:
      - name: Update PR
        uses: mscharley/dependency-changesets-action@648188b9f7d7cc54e59bcd367f99706325145257
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          use-conventional-commits: false
          commit-message: "chore(deps): changeset for dependency update [skip dependabot]"
          author-name: github-actions[bot]
          author-email: github-actions[bot]@users.noreply.github.com
```

## 2 - using Personal Access Token

This is a step that I personally don't like. If you repo are located in your personal account, this may seems a good option for you. 

I don't tested this approach, but I think you need a token with READ+WRITE permission in your repo.

Follow the github docs: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens

```yml
name: Add changeset to Dependabot updates

on:
  pull_request_target:
    types: [opened, synchronize, labeled]

jobs:
  dependabot:
    name: Update Dependabot PR
    if: contains(github.event.pull_request.labels.*.name, 'dependencies')

    runs-on: ubuntu-latest

    permissions:
      contents: write

    steps:
      - name: Update PR
        uses: mscharley/dependency-changesets-action@648188b9f7d7cc54e59bcd367f99706325145257
        with:
          token: ${{ secrets.CI_DEPENDABOT_CHANGESETS_PAT_TOKEN }}
          use-conventional-commits: false
          commit-message: "chore(deps): changeset for dependency update [skip dependabot]"
          author-name: github-actions[bot]
          author-email: github-actions[bot]@users.noreply.github.com
```

## 3 - using Github App Token

1 - Registering a new Github App

docs: https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/registering-a-github-app#registering-a-github-app

2 - use `actions/create-github-app-token`

docs: https://github.com/actions/create-github-app-token?tab=readme-ov-file#create-github-app-token

<details>

<summary>my how-to</summary>


tl;dr: go to one of the following links and click on the "New Github App"

- Personal scope: `https://github.com/settings/apps`
- Organization scope: `https://github.com/organizations/<org>/settings/apps`

-> provide a homepage url
-> you can disable the webhook part

-> permissions needed:

- Repository contents: Read & write

feel free to change other options, but it's not needed

after this, you can create the bot

---

after the creation, you will see the app management page

copy the App Id (keep this page opened)

in other tab, create a secret in your repo with someting like "CI_DEPENDABOT_CHANGESET_APP_ID" and paste the app id, then save

go back to app management page and create a new private key

back on the repo secrets page, create a secret like "CI_DEPENDABOT_CHANGESET_APP_PRIVATE_KEY", put the .pem file contents and save


---

after the app_id and app_private_key secrets setup in your target repo, go back to the app management page, and go to "install app" to install the app in your repo

</details> 

```yml
name: Add changeset to Dependabot updates

on:
  workflow_dispatch:
  pull_request_target:
    types: [opened, synchronize, labeled]

jobs:
  dependabot:
    name: Update Dependabot PR
    if: contains(github.event.pull_request.labels.*.name, 'dependencies')

    runs-on: ubuntu-latest

    permissions:
      contents: write

    steps:
      - uses: actions/create-github-app-token@v2
        id: app-token
        with:
          app-id: ${{ secrets.CI_DEPENDABOT_CHANGESETS_APP_ID }}
          private-key: ${{ secrets.CI_DEPENDABOT_CHANGESETS_APP_PRIVATE_KEY }}

      - name: Get GitHub App User ID
        id: get-user-id
        run: echo "user-id=$(gh api "/users/${{ steps.app-token.outputs.app-slug }}[bot]" --jq .id)" >> "$GITHUB_OUTPUT"
        env:
          GH_TOKEN: ${{ steps.app-token.outputs.token }}

      - name: Update PR
        uses: mscharley/dependency-changesets-action@648188b9f7d7cc54e59bcd367f99706325145257
        with:
          token: ${{ steps.app-token.outputs.token }}
          use-conventional-commits: false
          commit-message: "chore(deps): changeset for dependency update [skip dependabot]"
          author-name: ${{ steps.app-token.outputs.app-slug }}[bot]
          author-email: ${{ steps.get-user-id.outputs.user-id }}+${{ steps.app-token.outputs.app-slug }}[bot]@users.noreply.github.com
```

**Comment by guesant** [comment]
for who wonders about security like me, i made a environment to save the app id and app private key, that needs manual permission to run, and a ruleset to only allow the dependabot and my bot to manage the dependabot branches

![Image](https://github.com/user-attachments/assets/880f72b5-2b10-4aca-ab0a-7f08d872b08c)

![Image](https://github.com/user-attachments/assets/97fa9b3a-992b-49a7-8a3c-3190ebf67eb9)


the only issue i got about using `mscharley/dependency-changesets-action` is for dependabot pull requests that is not for package.json, in my case for dotnet.csproj: https://github.com/mscharley/dependency-changesets-action/issues/394

**Comment by mscharley** [comment]
@guessant The "new" fine grained PAT's can be limited to specific repositories when you create them, even for use with repos on a personal account and not an organisation account. I generally recommend people use a PAT for this action for simplicities sake. The only permission you need is "read/write content" (required since it adds a new commit to the PR branch), and any token that will grant that permission level should work fine. Any limitations on the account owning the PAT will also apply to the PAT itself, so if you generate it as a user with regular access to the repo then they will not be able to bypass any measures you have in place. This user can be a machine user if you want (for organisation repos).

For instance, this is a screenshot of the PAT I use for managing a few of my JavaScript libraries:

![Image](https://github.com/user-attachments/assets/26193f70-4663-4ef8-a4ec-0d2fadcb8375)

A Github app works around a few convenience issues, so I'm glad to see an example of this, that's pretty cool. But it shouldn't be any more secure, if you setup your PAT well.

**Comment by gillyb** [comment]
I really need this feature as well!... :)
I tried going ahead and implementing it, and created a PR for this: https://github.com/changesets/changesets/pull/1699

I saw that I didn't add enough test coverage, but I asked if this is something that will be accepted if I do.. I am willing to do the extra work to cover it, I just want to know first if this will be accepted if I do.

(I don't want to work hard if the maintainers don't think this is a good feature)

**Comment by IzumiSy** [comment]
A bit late to the party here, but I am still looking for the way to create changesets from dependabot updates.

I myself created [deps2changesets](https://github.com/IzumiSy/deps2changesets) that looks up the changes in package.json even on the monorepo and creates changesets automatically. It might be usable on Github Action. Hope it helps someone here.

**Comment by StyleShit** [comment]
I would love a CLI flag that would achieve this. In the meantime, it's doable by composing some of the changesets-provided utils:

https://github.com/drivenets/design-system/blob/e6aefa9a5/.github/workflows/dependabot-changeset.yml
https://github.com/drivenets/design-system/blob/e6aefa9a5/.github/scripts/generate-dependabot-changeset.ts

As mentioned above, using `pull_request_target` is a potential security issue, so I've added a check specifically for dependabot PRs
